source/http2_stream_manager.c

```diff
@@ -0,0 +1,1000 @@
+         */
+        struct aws_h2_sm_connection *chosen_connection =
+            s_get_best_sm_connection_from_set(&stream_manager->synced_data.ideal_available_set);
+        AWS_ASSERT(chosen_connection);
+        pending_stream_acquisition->sm_connection = chosen_connection;
+        chosen_connection->num_streams_assigned++;
+
+        STREAM_MANAGER_LOGF(
+            TRACE,
+            stream_manager,
+            "Picking connection:%p for acquisition:%p. Streams assigned to the connection=%" PRIu32 "",
+            (void *)chosen_connection->connection,
+            (void *)pending_stream_acquisition,
+            chosen_connection->num_streams_assigned);
+        /* Check if connection is still available or ideal, and move it if it's not */
+        if (chosen_connection->num_streams_assigned >= chosen_connection->max_concurrent_streams) {
+            /* It becomes not available for new streams any more, remove it from the set, but still alive (streams
+             * created will track the lifetime) */
+            chosen_connection->state = AWS_H2SMCST_FULL;
+            errored |=
+                aws_random_access_set_remove(&stream_manager->synced_data.ideal_available_set, chosen_connection);
+            STREAM_MANAGER_LOGF(
+                DEBUG,
+                stream_manager,
+                "connection:%p reaches max concurrent streams limits. Streams assigned to the connection=%" PRIu32 ""
+                "while limits=%" PRIu32 ". Moving it out of available connections.",
+                (void *)chosen_connection->connection,
+                chosen_connection->num_streams_assigned,
+                chosen_connection->max_concurrent_streams);
+        } else if (chosen_connection->num_streams_assigned >= stream_manager->ideal_concurrent_streams_per_connection) {
+            /* It meets the ideal limit, but still available for new streams, move it to the nonidea-available set */
+            errored |=
+                aws_random_access_set_remove(&stream_manager->synced_data.ideal_available_set, chosen_connection);
+            bool added = false;
+            errored |= aws_random_access_set_add(
+                &stream_manager->synced_data.nonideal_available_set, chosen_connection, &added);
+            errored |= !added;
+            chosen_connection->state = AWS_H2SMCST_NEARLY_FULL;
+            STREAM_MANAGER_LOGF(
+                DEBUG,
+                stream_manager,
+                "connection:%p reaches ideal concurrent streams limits. Ideal limits=%zu. Moving it to nonlimited set.",
+                (void *)chosen_connection->connection,
+                stream_manager->ideal_concurrent_streams_per_connection);
+        }
+    } else if (stream_manager->synced_data.holding_connections_count == stream_manager->max_connections) {
+        /**
+         * Try assigning to connection from nonideal available set. Only happens when we hold all the connections we can
+         * have
```

Comment:
comment why we do this. It took us a lot of design time to get right:

Note that we do not assign to nonideal connections until we're holding all the connections we can ever possibly get. This way, we don't overfill the first connections we get our hands on.